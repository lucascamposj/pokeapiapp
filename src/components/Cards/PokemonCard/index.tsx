import React, { useMemo } from 'react';
import { ScrollView, ActivityIndicator } from 'react-native';
import { DefaultTheme } from 'styled-components/native';

import TypesObject from '../../Sections/TypesSection/Type/types';

import BackIcon from '../../../assets/svg/icons/angle-left.svg';
import { Pokemon } from '../../../pages/Pokemon';
import Ruler from '../../../assets/svg/icons/ruler.svg';
import Weight from '../../../assets/svg/icons/weight.svg';
import TypesSection from '../../Sections/TypesSection';
import StatsSection from '../../Sections/StatsSection';
import AbilitiesSection from '../../Sections/AbilitiesSection';
import FormsSection from '../../Sections/FormsSection';

import {
  Container,
  MainCard,
  XPContainer,
  XPText,
  SpritesContainer,
  Sprite,
  SpriteSelector,
  NameContainer,
  NameText,
  IdText,
  HeightAndWeightContainer,
  HeightContainer,
  HeightText,
  WeightContainer,
  WeightText,
  SpeciesContainer,
  SpeciesText,
  Separator,
  SectionTitle,
  Header,
  HeaderBar,
  BackButton,
} from './styles';

interface PokemonCardProps {
  pokemon: Pokemon | undefined;
  loading: boolean;
  backButtonCallback: () => void;
  theme: DefaultTheme;
}

const getPokemonFirstType = (pokemon: Pokemon | undefined): string | null => {
  const firstType =
    pokemon?.types && pokemon.types.length > 0 ? pokemon.types[0] : null;
  const headerColor =
    firstType && firstType?.type.name
      ? TypesObject[firstType?.type.name].color
      : null;
  return headerColor;
};

const PokemonCard: React.FC<PokemonCardProps> = ({
  pokemon,
  loading,
  backButtonCallback,
  theme,
}: PokemonCardProps) => {
  const headerColor = useMemo(() => {
    if (pokemon && !loading) {
      const color = getPokemonFirstType(pokemon);
      return color || theme.colors.primary.main;
    }
    return theme.colors.common.white;
  }, [pokemon, loading, theme.colors]);

  const baseExperienceFormatted = useMemo(() => {
    return `${pokemon?.base_experience} XP`;
  }, [pokemon?.base_experience]);

  const idFormatted = useMemo(() => {
    return `#${pokemon?.id}`;
  }, [pokemon?.id]);

  const speciesFormatted = useMemo(() => {
    return `species: ${pokemon?.species.name}`;
  }, [pokemon?.species]);

  if (loading) {
    return <ActivityIndicator size="large" style={{ flex: 1 }} color="red" />;
  }

  return (
    <>
      <Header color={headerColor}>
        <HeaderBar>
          <BackButton onPress={backButtonCallback}>
            <BackIcon height={30} width={30} color="#FFF" />
          </BackButton>
        </HeaderBar>
        <SpritesContainer>
          <Sprite
            resizeMode="contain"
            source={{ uri: pokemon?.sprites.front_default }}
          />
          <SpriteSelector />
        </SpritesContainer>
      </Header>

      <ScrollView scrollIndicatorInsets={{ right: 1 }}>
        <Container>
          <MainCard>
            <XPContainer>
              <XPText>{baseExperienceFormatted}</XPText>
            </XPContainer>
            <NameContainer>
              <NameText>{pokemon?.name}</NameText>
              <IdText>{idFormatted}</IdText>
            </NameContainer>
            <HeightAndWeightContainer>
              <HeightContainer>
                <Ruler height={28} width={28} color="#333" />
                <HeightText>{pokemon?.height}</HeightText>
              </HeightContainer>
              <WeightContainer>
                <Weight height={23} width={23} color="#333" />
                <WeightText>{pokemon?.weight}</WeightText>
              </WeightContainer>
            </HeightAndWeightContainer>
            <SpeciesContainer>
              <SpeciesText>{speciesFormatted}</SpeciesText>
            </SpeciesContainer>
            <TypesSection types={pokemon?.types} />
            <Separator />
            <SectionTitle>Stats</SectionTitle>
            <StatsSection stats={pokemon?.stats} />
            <Separator />
            <SectionTitle>Abilities</SectionTitle>
            <AbilitiesSection abilities={pokemon?.abilities} />
            <Separator />
            <SectionTitle>Forms</SectionTitle>
            <FormsSection forms={pokemon?.forms} />
          </MainCard>
        </Container>
      </ScrollView>
    </>
  );
};

export default PokemonCard;
