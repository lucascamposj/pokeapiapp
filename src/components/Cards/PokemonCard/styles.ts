import styled from 'styled-components/native';
import { getStatusBarHeight } from 'react-native-iphone-x-helper';

export const Container = styled.View`
  flex: 1;
  width: 100%;
  align-content: center;
  padding: 32px 24px 16px;
`;
export const MainCard = styled.View`
  width: 100%;
`;
export const XPContainer = styled.View``;
export const XPText = styled.Text`
  font-size: 25px;
  font-family: 'RobotoMono-Light';
`;
export const SpritesContainer = styled.View`
  align-items: center;
  width: 100%;
  height: 200px;
  margin: 0 auto;
  margin-top: -40px;
  z-index: 998;
  elevation: 18;
`;

export const Sprite = styled.Image`
  height: 100%;
  width: 100%;
`;
export const SpriteSelector = styled.View``;
export const NameContainer = styled.View`
  align-items: center;
  justify-content: center;
  flex-direction: row;
`;
export const NameText = styled.Text`
  margin-right: 5px;
  font-size: 30px;
  text-transform: capitalize;
  font-family: 'RobotoMono-Bold';
`;
export const IdText = styled.Text`
  font-size: 15px;
  color: #999;
`;
export const HeightAndWeightContainer = styled.View`
  padding: 20px 20px;
  align-items: center;
  justify-content: center;
  flex-direction: row;
`;
export const HeightContainer = styled.View`
  flex: 1;
  align-items: center;
  justify-content: center;
  flex-direction: row;
`;
export const HeightText = styled.Text`
  font-size: 20px;
  margin-left: 10px;
  font-family: 'RobotoMono-Light';
`;

export const WeightContainer = styled.View`
  flex: 1;
  align-items: center;
  justify-content: center;
  flex-direction: row;
`;
export const WeightText = styled.Text`
  font-size: 20px;
  margin-left: 10px;
  font-family: 'RobotoMono-Light';
`;
export const SpeciesContainer = styled.View`
  align-items: center;
  justify-content: center;
`;
export const SpeciesText = styled.Text``;
export const TypesContainer = styled.View``;
export const Type = styled.View``;
export const Stats = styled.View``;
export const Separator = styled.View`
  background-color: #bbb;
  height: 2px;
  width: 100%;
  margin: 10px 0;
`;
export const SectionTitle = styled.Text`
  text-align: center;
  font-family: 'RobotoMono-Bold';
  font-size: 18px;
`;
export const Abilities = styled.View``;
export const Forms = styled.View``;

interface HeaderProps {
  color: string;
}
export const Header = styled.View<HeaderProps>`
  padding: 24px;
  padding-top: ${getStatusBarHeight() + 24}px;
  background-color: ${({ color, theme }) => color || theme.colors.common.white};
  height: 170px;
  z-index: 998;
  align-items: flex-start;
  shadow-color: ${({ color, theme }) => color || theme.colors.common.white};
  shadow-offset: 0 10px;
  shadow-opacity: 0.9;
  shadow-radius: 18px;

  elevation: 17;
`;

export const HeaderBar = styled.View`
  width: 100%;
  z-index: 999;
`;

export const HeaderTitle = styled.Text`
  color: ${({ theme }) => theme.colors.primary.contrastText};
  font-size: 24px;
  font-weight: bold;
  font-family: 'RobotoMono-Regular';
  line-height: 28px;
  text-transform: capitalize;
`;

export const BackButton = styled.TouchableOpacity`
  margin-right: 15px;
  z-index: 999;

  elevation: 20;
`;
