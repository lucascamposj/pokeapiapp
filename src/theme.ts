// theme.ts
import { DefaultTheme } from 'styled-components/native';

export const defaultTheme: DefaultTheme = {
  colors: {
    common: {
      red: '#f00',
      white: '#ffffff',
      gray: '#333',
      black: '#000',
    },
    primary: {
      main: '#f00',
      contrastText: '#fff',
    },
    secondary: {
      main: '#fff',
      contrastText: '#000',
    },
  },
};
