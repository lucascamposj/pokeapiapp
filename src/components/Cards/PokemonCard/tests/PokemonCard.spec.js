import React from 'react';
import PokemonCard from '../../PokemonCard';
import {shallow} from 'enzyme';
import { defaultTheme } from '../../../../theme';

import mockedPokemon from '../tests/__mocks__/mockedPokemon.json';
import mockedPokemonWithoutType from '../tests/__mocks__/mockedPokemonWithoutType.json';

describe('PokemonCard', () => {
    describe('Rendering', () => {
        it('should match to snapshot loading', () => {
            const component = shallow(
              <PokemonCard
                pokemon={mockedPokemon}
                loading={true}
                theme={defaultTheme}
              />
            )
            expect(component).toMatchSnapshot()
        });

        it('should match to snapshot loaded', () => {
          const component = shallow(
            <PokemonCard
              pokemon={mockedPokemon}
              loading={false}
              theme={defaultTheme}
            />
          )
          expect(component).toMatchSnapshot()
        });

        it('should match to snapshot with Pokemon without a type', () => {
          const component = shallow(
            <PokemonCard
              pokemon={mockedPokemonWithoutType}
              loading={false}
              theme={defaultTheme}
            />
          )
          expect(component).toMatchSnapshot()
        });
    });
});
