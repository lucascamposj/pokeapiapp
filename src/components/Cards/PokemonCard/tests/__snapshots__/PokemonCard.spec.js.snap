// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonCard Rendering should match to snapshot loaded 1`] = `
<Fragment>
  <Styled(View)
    color="#5DBE62"
  >
    <Styled(View)>
      <Styled(Component)>
        <SvgMock
          color="#FFF"
          height={30}
          width={30}
        />
      </Styled(Component)>
    </Styled(View)>
    <Styled(View)>
      <Styled(Image)
        resizeMode="contain"
        source={
          Object {
            "uri": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
          }
        }
      />
      <Styled(View) />
    </Styled(View)>
  </Styled(View)>
  <ScrollView
    scrollIndicatorInsets={
      Object {
        "right": 1,
      }
    }
  >
    <Styled(View)>
      <Styled(View)>
        <Styled(View)>
          <Styled(Text)>
            64 XP
          </Styled(Text)>
        </Styled(View)>
        <Styled(View)>
          <Styled(Text)>
            bulbasaur
          </Styled(Text)>
          <Styled(Text)>
            #1
          </Styled(Text)>
        </Styled(View)>
        <Styled(View)>
          <Styled(View)>
            <SvgMock
              color="#333"
              height={28}
              width={28}
            />
            <Styled(Text)>
              7
            </Styled(Text)>
          </Styled(View)>
          <Styled(View)>
            <SvgMock
              color="#333"
              height={23}
              width={23}
            />
            <Styled(Text)>
              69
            </Styled(Text)>
          </Styled(View)>
        </Styled(View)>
        <Styled(View)>
          <Styled(Text)>
            species: bulbasaur
          </Styled(Text)>
        </Styled(View)>
        <TypeSection
          types={
            Array [
              Object {
                "slot": 1,
                "type": Object {
                  "name": "grass",
                  "url": "https://pokeapi.co/api/v2/type/12/",
                },
              },
              Object {
                "slot": 2,
                "type": Object {
                  "name": "poison",
                  "url": "https://pokeapi.co/api/v2/type/4/",
                },
              },
            ]
          }
        />
        <Styled(View) />
        <Styled(Text)>
          Stats
        </Styled(Text)>
        <StatSection
          stats={
            Array [
              Object {
                "base_stat": 45,
                "effort": 0,
                "stat": Object {
                  "name": "hp",
                  "url": "https://pokeapi.co/api/v2/stat/1/",
                },
              },
              Object {
                "base_stat": 49,
                "effort": 0,
                "stat": Object {
                  "name": "attack",
                  "url": "https://pokeapi.co/api/v2/stat/2/",
                },
              },
              Object {
                "base_stat": 49,
                "effort": 0,
                "stat": Object {
                  "name": "defense",
                  "url": "https://pokeapi.co/api/v2/stat/3/",
                },
              },
              Object {
                "base_stat": 65,
                "effort": 1,
                "stat": Object {
                  "name": "special-attack",
                  "url": "https://pokeapi.co/api/v2/stat/4/",
                },
              },
              Object {
                "base_stat": 65,
                "effort": 0,
                "stat": Object {
                  "name": "special-defense",
                  "url": "https://pokeapi.co/api/v2/stat/5/",
                },
              },
              Object {
                "base_stat": 45,
                "effort": 0,
                "stat": Object {
                  "name": "speed",
                  "url": "https://pokeapi.co/api/v2/stat/6/",
                },
              },
            ]
          }
        />
        <Styled(View) />
        <Styled(Text)>
          Abilities
        </Styled(Text)>
        <AbilitySection
          abilities={
            Array [
              Object {
                "ability": Object {
                  "name": "overgrow",
                  "url": "https://pokeapi.co/api/v2/ability/65/",
                },
                "is_hidden": false,
                "slot": 1,
              },
              Object {
                "ability": Object {
                  "name": "chlorophyll",
                  "url": "https://pokeapi.co/api/v2/ability/34/",
                },
                "is_hidden": true,
                "slot": 3,
              },
            ]
          }
        />
        <Styled(View) />
        <Styled(Text)>
          Forms
        </Styled(Text)>
        <FormSection
          forms={
            Array [
              Object {
                "name": "bulbasaur",
                "url": "https://pokeapi.co/api/v2/pokemon-form/1/",
              },
            ]
          }
        />
      </Styled(View)>
    </Styled(View)>
  </ScrollView>
</Fragment>
`;

exports[`PokemonCard Rendering should match to snapshot loading 1`] = `
<ActivityIndicator
  animating={true}
  color="red"
  hidesWhenStopped={true}
  size="large"
  style={
    Object {
      "flex": 1,
    }
  }
/>
`;

exports[`PokemonCard Rendering should match to snapshot with Pokemon without a type 1`] = `
<Fragment>
  <Styled(View)
    color="#f00"
  >
    <Styled(View)>
      <Styled(Component)>
        <SvgMock
          color="#FFF"
          height={30}
          width={30}
        />
      </Styled(Component)>
    </Styled(View)>
    <Styled(View)>
      <Styled(Image)
        resizeMode="contain"
        source={
          Object {
            "uri": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
          }
        }
      />
      <Styled(View) />
    </Styled(View)>
  </Styled(View)>
  <ScrollView
    scrollIndicatorInsets={
      Object {
        "right": 1,
      }
    }
  >
    <Styled(View)>
      <Styled(View)>
        <Styled(View)>
          <Styled(Text)>
            64 XP
          </Styled(Text)>
        </Styled(View)>
        <Styled(View)>
          <Styled(Text)>
            bulbasaur
          </Styled(Text)>
          <Styled(Text)>
            #1
          </Styled(Text)>
        </Styled(View)>
        <Styled(View)>
          <Styled(View)>
            <SvgMock
              color="#333"
              height={28}
              width={28}
            />
            <Styled(Text)>
              7
            </Styled(Text)>
          </Styled(View)>
          <Styled(View)>
            <SvgMock
              color="#333"
              height={23}
              width={23}
            />
            <Styled(Text)>
              69
            </Styled(Text)>
          </Styled(View)>
        </Styled(View)>
        <Styled(View)>
          <Styled(Text)>
            species: bulbasaur
          </Styled(Text)>
        </Styled(View)>
        <TypeSection
          types={Array []}
        />
        <Styled(View) />
        <Styled(Text)>
          Stats
        </Styled(Text)>
        <StatSection
          stats={
            Array [
              Object {
                "base_stat": 45,
                "effort": 0,
                "stat": Object {
                  "name": "hp",
                  "url": "https://pokeapi.co/api/v2/stat/1/",
                },
              },
              Object {
                "base_stat": 49,
                "effort": 0,
                "stat": Object {
                  "name": "attack",
                  "url": "https://pokeapi.co/api/v2/stat/2/",
                },
              },
              Object {
                "base_stat": 49,
                "effort": 0,
                "stat": Object {
                  "name": "defense",
                  "url": "https://pokeapi.co/api/v2/stat/3/",
                },
              },
              Object {
                "base_stat": 65,
                "effort": 1,
                "stat": Object {
                  "name": "special-attack",
                  "url": "https://pokeapi.co/api/v2/stat/4/",
                },
              },
              Object {
                "base_stat": 65,
                "effort": 0,
                "stat": Object {
                  "name": "special-defense",
                  "url": "https://pokeapi.co/api/v2/stat/5/",
                },
              },
              Object {
                "base_stat": 45,
                "effort": 0,
                "stat": Object {
                  "name": "speed",
                  "url": "https://pokeapi.co/api/v2/stat/6/",
                },
              },
            ]
          }
        />
        <Styled(View) />
        <Styled(Text)>
          Abilities
        </Styled(Text)>
        <AbilitySection
          abilities={
            Array [
              Object {
                "ability": Object {
                  "name": "overgrow",
                  "url": "https://pokeapi.co/api/v2/ability/65/",
                },
                "is_hidden": false,
                "slot": 1,
              },
              Object {
                "ability": Object {
                  "name": "chlorophyll",
                  "url": "https://pokeapi.co/api/v2/ability/34/",
                },
                "is_hidden": true,
                "slot": 3,
              },
            ]
          }
        />
        <Styled(View) />
        <Styled(Text)>
          Forms
        </Styled(Text)>
        <FormSection
          forms={
            Array [
              Object {
                "name": "bulbasaur",
                "url": "https://pokeapi.co/api/v2/pokemon-form/1/",
              },
            ]
          }
        />
      </Styled(View)>
    </Styled(View)>
  </ScrollView>
</Fragment>
`;
